tensorflow/python/keras/layers/wrappers.py

```diff
@@ -148,56 +150,73 @@ def _get_shape_tuple(self, init_tuple, tensor, start_idx, int_shape=None):
-    if len(input_shape) < 3:
+    input_shape = tf_utils.convert_shapes(input_shape, to_tuples=False)
+    input_dims = generic_utils.to_list(nest.map_structure(lambda x: x.ndims, input_shape))
+    if all(dim < 3 for dim in input_dims):
       raise ValueError(
           '`TimeDistributed` Layer should be passed an `input_shape ` '
           'with at least 3 dimensions, received: ' + str(input_shape))
     # Don't enforce the batch or time dimension.
-    self.input_spec = InputSpec(shape=[None, None] + input_shape[2:])
-    child_input_shape = [input_shape[0]] + input_shape[2:]
+    self.input_spec = nest.map_structure(lambda x: InputSpec(shape=[None, None] + x.as_list()[2:]),
+                                         input_shape)
+    child_input_shape = nest.map_structure(
+        lambda x: tensor_shape.TensorShape([x.as_list()[0]] + x.as_list()[2:]),
+        input_shape)
+    child_input_shape = tf_utils.convert_shapes(child_input_shape)
     super(TimeDistributed, self).build(tuple(child_input_shape))
     self.built = True
 
   def compute_output_shape(self, input_shape):
-    input_shape = tensor_shape.TensorShape(input_shape).as_list()
-    child_input_shape = tensor_shape.TensorShape([input_shape[0]] +
-                                                 input_shape[2:])
+    input_shape = tf_utils.convert_shapes(input_shape, to_tuples=False)
+    def remove_timesteps(dims):
+      dims = dims.as_list()
+      return tensor_shape.TensorShape([dims[0]] + dims[2:])
+
+    child_input_shape = nest.map_structure(remove_timesteps, input_shape)
     child_output_shape = self.layer.compute_output_shape(child_input_shape)
-    if not isinstance(child_output_shape, tensor_shape.TensorShape):
-      child_output_shape = tensor_shape.TensorShape(child_output_shape)
-    child_output_shape = child_output_shape.as_list()
-    timesteps = input_shape[1]
-    return tensor_shape.TensorShape([child_output_shape[0], timesteps] +
-                                    child_output_shape[1:])
+    child_output_shape = tf_utils.convert_shapes(child_output_shape,
+                                                 to_tuples=False)
+    timesteps = tf_utils.convert_shapes(input_shape)
+    timesteps = nest.flatten(timesteps)[1]
+    def insert_timesteps(dims):
+      dims = dims.as_list()
+      return tensor_shape.TensorShape([dims[0], timesteps] + dims[1:])
+
+    return nest.map_structure(insert_timesteps, child_output_shape)
 
   def call(self, inputs, training=None, mask=None):
     kwargs = {}
     if generic_utils.has_arg(self.layer.call, 'training'):
       kwargs['training'] = training
 
-    input_shape = K.int_shape(inputs)
-    if input_shape[0] and not self._always_use_reshape:
+    input_shape = nest.map_structure(
+        lambda x: tensor_shape.TensorShape(K.int_shape(x)),
+        inputs)
+    batch_size = tf_utils.convert_shapes(input_shape)
+    batch_size = nest.flatten(batch_size)[0]
```

Comment:
Same here, you probably should get the first shape from the flattened list, and then get the first dim.

Reply:
See comment above.